src/components/chat-components/ChatSingleMessage.tsx

```diff
@@ -79,6 +85,100 @@ const ChatSingleMessage: React.FC<ChatSingleMessageProps> = ({
     });
   };
 
+  const handleApplyCode = useCallback(
```

Comment:
I would also suggest create a custom hook and move this chunk of code to a separate module under composer. Making composer code more isolated from the core chat.